storm-client/src/jvm/org/apache/storm/messaging/netty/KerberosSaslClientHandler.java

```diff
@@ -96,53 +96,69 @@ private void handleControlMessage(ChannelHandlerContext ctx, ControlMessage cont
-            byte[] responseToServer = saslNettyClient
-            .saslResponse(saslMessageToken);
-            if (responseToServer == null) {
-                // If we generate a null response, then authentication has completed
-                // (if not, warn), and return without sending a response back to the
-                // server.
-                LOG.debug("Response to server is null: authentication should now be complete.");
-                if (!saslNettyClient.isComplete()) {
-                    LOG.warn("Generated a null response, but authentication is not complete.");
-                    throw new Exception("Our reponse to the server is null, but as far as we can tell, we are not authenticated yet.");
-                }
-            this.client.channelReady(channel);
-            } else {
-                LOG.debug("Response to server token has length: {}",
-                          responseToServer.length);
-            // Construct a message containing the SASL response and send it to the
+        LOG.debug("Responding to server's token of length: {}", saslMessageToken.getSaslToken().length);
+
+        // Generate SASL response (but we only actually send the response if
+        // it's non-null.
+        byte[] responseToServer = saslNettyClient.saslResponse(saslMessageToken);
+        if (responseToServer == null) {
+            // If we generate a null response, then authentication has completed
+            // (if not, warn), and return without sending a response back to the
             // server.
+            LOG.debug("Response to server is null: authentication should now be complete.");
+            if (!saslNettyClient.isComplete()) {
+                LOG.warn("Generated a null response, but authentication is not complete.");
+                throw new Exception("Our reponse to the server is null, but as far as we can tell, we are not authenticated yet.");
+            }
+            this.client.channelReady(channel);
+        } else {
+            LOG.debug("Response to server token has length: {}",
+                      responseToServer.length);
+            // Construct a message containing the SASL response and send it to the server.
             SaslMessageToken saslResponse = new SaslMessageToken(responseToServer);
             channel.writeAndFlush(saslResponse, channel.voidPromise());
         }
     }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        super.channelRegistered(ctx);
+        LOG.debug("channelRegistered {}", ctx);
+    }
+
+    @Override
+    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+        super.channelUnregistered(ctx);
+        LOG.debug("channelUnregistered {}", ctx);
+
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+        LOG.warn("{} exceptionCaught", ctx, cause);
+        super.exceptionCaught(ctx, cause);
```

Comment:
LOG.warn("{} exceptionCaught", ctx, cause) you have one {} but 2 variables here.

Reply:
@danny0405 
Please refer https://www.slf4j.org/faq.html#paramException to see what slf4j2 does the magic.